### PR TITLE
Afad connector

### PR DESCRIPTION
**Pull request by Azomasiel**
Fixes #

## Proposed Changes

Update data parser to support a new AFAD feature
Add instruction to install the Syslog proxy using the intended Quickstart template.

### CODE REVIEW

**Comment by preetikr** [comment]
@v-jayakal - Please review the data connector updates
@shainw  - please review the parser changes for Alsid.

Thanks!

**Comment by v-admahe** [comment]
Could you please change the legend to **"AlsidForADLog_CL"** as legend should be datatype name.

Thanks,
Aditi Maheshwari

**Comment by v-admahe** [comment]
@Azomasiel Could you please change the legend to "AlsidForADLog_CL" as legend should be datatype name.

Thanks,
Aditi Maheshwari

**Comment by v-rucdu** [comment]
> Could you please change the legend to **"AlsidForADLog_CL"** as legend should be datatype name.
> 
> Thanks,
> Aditi Maheshwari

@Azomasiel : Please resolve the review comments.... Thanks!!!

**Comment by Aorimn** [comment]
Sorry for the delay, the review comment should be fixed now

**Comment by v-admahe** [comment]
@Azomasiel : Permission tag is not matching with the sample template due to this test cases are failing. you can refer the sample template from here : https://github.com/Azure/Azure-Sentinel/blob/master/DataConnectors/Templates/Connector_REST_API_template.json

or you can use this text:

"permissions": {
        "resourceProvider": [
            {
                "provider": "Microsoft.OperationalInsights/workspaces",
                "permissionsDisplayText": "read and write permissions are required.",
                "providerDisplayName": "Workspace",
                "scope": "Workspace",
                "requiredPermissions": {
                    "write": true,
                    "read": true,
                    "delete": true
                }
            },
            {
                "provider": "Microsoft.OperationalInsights/workspaces/sharedKeys",
                "permissionsDisplayText": "read permissions to shared keys for the workspace are required. [See the documentation to learn more about workspace keys](https://docs.microsoft.com/azure/azure-monitor/platform/agent-windows#obtain-workspace-id-and-key).",
                "providerDisplayName": "Keys",
                "scope": "Workspace",
                "requiredPermissions": {
                    "action": true
                }
            }
        ],

Thanks,
Aditi Maheshwari

**Comment by v-admahe** [comment]
@Azomasiel  :Test cases are failing due to dot is missing at the end of permission display text so could you please add full stop at the end of the sentence.

Use this line: read permissions to shared keys for the workspace are required. See the documentation to learn more about workspace keys.

**Comment by Aorimn** [comment]
Thanks for the insights @v-admahe.

Should we also update https://github.com/Azure/Azure-Sentinel/blob/master/DataConnectors/Templates/Connector_REST_API_template.json#L52 to add this ending dot? I literally copied/pasted the permissions block from here, hence the missing dot.

**Comment by v-admahe** [comment]
> Thanks for the insights @v-admahe.
> 
> Should we also update https://github.com/Azure/Azure-Sentinel/blob/master/DataConnectors/Templates/Connector_REST_API_template.json#L52 to add this ending dot? I literally copied/pasted the permissions block from here, hence the missing dot.

We already raise the request for this once approved, it will reflected in the git hub.

**Comment by v-admahe** [comment]
@Azomasiel : There is merge conflict in the PR could you please check it once and resolve it.

Thanks.
Aditi Maheshwari

**Review by v-admahe — DISMISSED** [review]

**Review by v-jayakal — APPROVED** [review]

**Comment by v-jayakal** [comment]
@shainw, Can you please review the Parser and detections please?

File: Parsers/Alsid/afad_parser.kql [hunk 1]
```diff
@@ -91,6 +91,16 @@ let Changes = Common
                      EventID:string "\" \""
                      EventType:string "\" "
                      Attributes:string "\r?\n";
-union Changes, Deviances
+let Attacks = Common
```
Comment (shainw) [inline Parsers/Alsid/afad_parser.kql:94]: There is no MessageType = 2 in the [Alsid sample data set](https://github.com/Azure/Azure-Sentinel/blob/master/Sample%20Data/Syslog/AlsidForADLog_CL.json), so I could not test that specifically.

Also, I think the overall parser isn't working for after you do the initial parse on RawData.  DistinctPart doesn't appear to have the strings you expect as the \" all get removed in my testing.  This means this subsequent parsing doesn't seem to work.  I changed the Deviances and below portions to the following and it seems to work properly:
```
let Deviances = Common
| where MessageType == 0 
| parse DistinctPart with '"'
                     Codename:string '" "'
                     Severity:string '" "'
                     ADObject:string '" "'
                     DevianceID:string '" "'
                     ProfileID:string '" "'
                     ReasonCodename:string '" "'
                     EventID:string '"'
                     Attributes:string
                     ;
let Changes = Common
| where MessageType == 1
| parse DistinctPart with '"'
                     ADObject:string '" "'
                     EventID:string '" "'
                     EventType:string '" '
                     Attributes:string
                     ;
let Attacks = Common
| where MessageType == 2
| parse DistinctPart with '"'
                     Codename:string '" "'
                     Severity:string '" "'
                     SourceHostname:string '" "'
                     SourceIP:string '" "'
                     DestinationHostname:string '" "'
                     DestinationIP:string '" "'
                     Attributes:string
                     ;
```

**Review by shainw — CHANGES_REQUESTED** [review]

File: Parsers/Alsid/afad_parser.kql (hunk 1 above)
Comment (Azomasiel) [reply Parsers/Alsid/afad_parser.kql:94]: It used to work, in fact the version on master doesn't have the fix you propose. has anything changed to break the parser?

My internship at Alsid has ended so I do not have access to my Azure subscription anymore, is there any way I could test the parser myself with the sample data?

Comment (Aorimn) [reply Parsers/Alsid/afad_parser.kql:94]: @shainw thanks for the feedback. I've added some examples in the mentioned files, is that better now?

Comment (shainw) [reply Parsers/Alsid/afad_parser.kql:94]: @Aorimn - Thanks for adding the additional message type.  I am still seeing the parser not handle the parsing correctly. 
image

It looks to me like the escaping of the double quote is not being handled properly.  When I change to this, it works:

```
| parse RawData with
                     Time:datetime ' '
                     Host:string ' '
                     Product:string '['
                     PID:int ']: \"'
                     MessageType:int '\" \"'
                     AlertID:int '\" \"'
                     Forest:string '\" \"'
                     Domain:string '\" '
                     DistinctPart:string;
let Deviances = Common
| where MessageType == 0 
| parse DistinctPart with '"'
                     Codename:string '" "'
                     Severity:string '" "'
                     ADObject:string '" "'
                     DevianceID:string '" "'
                     ProfileID:string '" "'
                     ReasonCodename:string '" "'
                     EventID:string '"'
                     Attributes:string
                     ;
let Changes = Common
| where MessageType == 1
| parse DistinctPart with '"'
                     ADObject:string '" "'
                     EventID:string '" "'
                     EventType:string '" '
                     Attributes:string
                     ;
let Attacks = Common
| where MessageType == 2
| parse DistinctPart with '"'
                     Codename:string '" "'
                     Severity:string '" "'
                     SourceHostname:string '" "'
                     SourceIP:string '" "'
                     DestinationHostname:string '" "'
                     DestinationIP:string '" "'
                     Attributes:string
                     ;
```

**Review by shainw — CHANGES_REQUESTED** [review]

**Comment by fr4nco1s** [comment]
Hi @shainw, I see that you removed the "\r\n" in the parsing requests for Deviances/Changes/Attacks to get it working.
It looks like you RawData strings miss the \r\n at the end, and that explains why the parsing of the Deviance/Change/Attack does not work. How are your Alsid events sent to Azure Sentinel ? Do you use the OMS agent ?

**Comment by Azomasiel** [comment]
@fr4nco1s We do use the OMS agent, I guess the OMS agent always send the \r\n but I failed when I dumped the data into the sample data json.

**Comment by fr4nco1s** [comment]
Thank you @Azomasiel  that makes sens then.

If I correctly understand, the parsing does not work for the samples as they do not have '\r\n' at the end, but for real events it works fine for eveyone, including @shainw  ?

Now let's say that the change proposed by @shainw is accepted and the '\r\n' are removed from the parsing requests. It should not have any impact ? I mean, it will just add a '\r\n' at the end of the Attributes field, but it does not impact the good working of the Alsid workbooks etc... ?

**Comment by Azomasiel** [comment]
@fr4nco1s Exactly, it will not impact the connector since we do not use the Attributes field,
I initially put the \r\n to mark the end of line delimiter and prevent malformed input.

**Comment by shainw** [comment]
> @fr4nco1s Exactly, it will not impact the connector since we do not use the Attributes field,
> I initially put the \r\n to mark the end of line delimiter and prevent malformed input.

@Azomasiel - I need you to merge master back in and then we can complete or take the PR I submitted to your fork - https://github.com/AlsidOfficial/Azure-Sentinel/pull/4

**Comment by Azomasiel** [comment]
Thank you @shainw !

@Aorimn would it be possible whenever you can to merge [this PR](https://github.com/AlsidOfficial/Azure-Sentinel/pull/5) to merge master and then [this one](https://github.com/AlsidOfficial/Azure-Sentinel/pull/4)?

**Comment by fr4nco1s** [comment]
> > @fr4nco1s Exactly, it will not impact the connector since we do not use the Attributes field,
> > I initially put the \r\n to mark the end of line delimiter and prevent malformed input.
> 
> @Azomasiel - I need you to merge master back in and then we can complete or take the PR I submitted to your fork - [AlsidOfficial#4](https://github.com/AlsidOfficial/Azure-Sentinel/pull/4)

Hi @shainw your PR has been merged into the fork.

**Review by shainw — APPROVED** [review]

**Pull request merged**